eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs

```diff
@@ -163,6 +164,11 @@ data TxOut era
       {-# UNPACK #-} !(CompactAddr (Crypto era))
       !(CompactForm (Core.Value era))
       {-# UNPACK #-} !(BinaryData era) -- Inline data
+  | TxOutCompactRefScript'
+      {-# UNPACK #-} !(CompactAddr (Crypto era))
+      !(CompactForm (Core.Value era))
+      !(Datum era)
```

Comment:
this is kind of a compromise between performance and having a nice API + not polluting the CBOR with a bunch of extra constructors to represent every possible new permutation of `Addr × Value × (Datum | DataHash)? × Script?`

Reply:
It is not compromise between performance. It is compromise between memory overhead. `Datum era` takes up more than twice what the `BinaryData era` does!

I am afraid that in that we don't get UTxO HD in time we will need to save every byte possible. In my opinion, once we don't have to worry about memory overhead that much we can figure out how we can dampen this, but until then we should be careful!

Reply:
My previous comment was wrong. This looks good, I like it too! :wink:

Reply:
Does this mean that inline data and ref script can't be used in the same output?

Reply:
On the contrary, it means that inline datum is optional with a ref script

Reply:
`TxOutCompactRefScript' caddr value (Datum inlineDatum) refScript` is valid